ang/prospect/services/view-pledge-contribution-case-action.service.js

```diff
@@ -0,0 +1,107 @@
+(function (angular, $, _) {
+  var module = angular.module('prospect');
+
+  module.service('ViewPledgeContributionCaseAction', ViewPledgeContributionCaseAction);
+
+  function ViewPledgeContributionCaseAction ($location, crmApi, ProspectGlobalValues) {
+    var prospectConverted = {
+      isConverted: false,
+      paymentInfo: {
+        payment_completed: false,
+        pledge_balance: false,
+        payment_url: false,
+        payment_entity: false
+      }
+    };
+
+    /**
+     * Refresh Data for the Service
+     *
+     * @param {array} cases
+     */
+    this.refreshData = function (cases) {
+      if (cases[0]) {
+        var caseID = cases[0].id;
+
+        checkIfConvertedToProspect(caseID)
+          .then(function () {
+            getPaymentInfo(caseID);
+          });
+      }
+    };
+
+    /**
+     * Checks if the Action is allowed
+     *
+     * @param {Object} action
+     * @param {Array} cases
+     * @return {boolean}
+     */
+    this.isActionAllowed = function (action, cases) {
+      var isActionAllowed = action.type === 'pledge'
+        ? prospectConverted.paymentInfo.payment_entity === 'pledge'
+        : prospectConverted.paymentInfo.payment_entity === 'contribute';
+
+      return checkIfProspectingCaseTypeCategory() &&
+        prospectConverted.isConverted && isActionAllowed;
+    };
+
+    /**
+     * Click event handler for the Action
+     *
+     * @param {Array} cases
+     * @param {Object} action
+     * @param {Function} callbackFn
+     */
+    this.doAction = function (cases, action, callbackFn) {
+      var contactID = cases[0].client[0].contact_id;
+      var url = prospectConverted.paymentInfo.payment_url + '&cid=' + contactID;
+
+      return url;
+    };
+
+    /**
+     * Checks if the Case is Converted to Prospect
+     *
+     * @param {String/Int} caseID
+     * @return {Promise}
+     */
+    function checkIfConvertedToProspect (caseID) {
+      return crmApi('ProspectConverted', 'get', {
+        'sequential': 1,
+        'prospect_case_id': caseID
+      }).then(function (caseData) {
+        prospectConverted.isConverted = caseData.count > 0;
+      });
+    }
+
+    /**
+     * Gets the payment information
+     *
+     * @param {String/Int} caseID
+     */
+    function getPaymentInfo (caseID) {
+      if (prospectConverted.isConverted) {
+        crmApi('ProspectConverted', 'getpaymentinfo', {
+          'prospect_case_id': caseID
+        }).then(function (paymentInfo) {
+          prospectConverted.paymentInfo = paymentInfo;
+        });
+      }
```

Comment:
why do we need an `if` here? I find some issues with this function in general:

- it's called `getPaymentInfo`, but doesn't return anything.
- it might or might not "get" what I want depending on a condition I'm not aware of (`if (prospectConverted.isConverted) {`. What I mean is, anyone calling `getPaymentInfo` will not know why they are not getting the payment info if `prospectConverted.isConverted` is set to false.
- It's overwriting a global value `prospectConverted.paymentInfo`, so it's a setter, not a getter.

Reply:
Moved the setter functionality and the if condition outside.